juju/bundle.py

```diff
@@ -422,15 +565,47 @@ async def run(self, context):
         :param context: is used for any methods or properties required to
             perform a change.
         """
+
         # We don't add local charms because they've already been added
         # by self._handle_local_charms
-        if self.charm.startswith('local:'):
+        url = URL.parse(str(self.charm))
+        ch = None
+        identifier = None
+        if Schema.LOCAL.matches(url.schema):
+            origin = client.CharmOrigin(source="local", risk="stable")
+            context.origins[self.charm] = {str(None): origin}
             return self.charm
 
-        entity_id = await context.charmstore.entityId(self.charm)
-        log.debug('Adding %s', entity_id)
-        await context.client_facade.AddCharm(channel=None, url=entity_id, force=False)
-        return entity_id
+        elif Schema.CHARM_STORE.matches(url.schema):
```

Comment:
nitpick: these elifs could just be ifs. If .matches would match for multiple branches of the if, else if statements, then we have a bug, anyway.